rfcs/0004-npm-scripts.md

```diff
@@ -0,0 +1,133 @@
+by editing the `package.json` and copying the command from the custom event to
+a `preinstall` or `postinstall` script event that [npm and yarn recognizes](https://docs.npmjs.com/cli/v6/using-npm/scripts#npm-install)
+as part of the `npm install`.  Many users do not wish to make changes to their
+app source code, and there has been requests to support execution of these
+commands by the buildpacks
+([e.g.](https://github.com/paketo-buildpacks/yarn/issues/59)).
+
+## Implementation
+The proposed environment variables are as follows:
+
+#### BP_NPM_PREINSTALL
+```shell
+$BP_NPM_PREINSTALL="build,custom-script"
+```
+A comma separated list of life cycle events listed in the app's `package.json`
+that the `npm-install` buildpack must execute **before** it installs modules
+using `npm`.
+
+#### BP_NPM_POSTINSTALL
+```shell
+$BP_NPM_POSTINSTALL="build,custom-script"
+```
+A comma separated list of life cycle events listed in the app's `package.json`
+that the `npm-install` buildpack must execute **after** it installs modules using `npm`.
+
+#### BP_YARN_PREINSTALL
+```shell
+$BP_YARN_PREINSTALL="build,custom-script"
+```
+A comma separated list of life cycle events listed in the app's `package.json`
+that the `yarn-install` buildpack must execute **before** it installs modules
+using `yarn`.
+
+#### BP_YARN_POSTINSTALL
+```shell
+$BP_YARN_POSTINSTALL="build,custom-script"
+```
+A comma separated list of life cycle events listed in the app's `package.json`
+that the `yarn-install` buildpack must execute **after** it installs modules
+using `yarn`.
+
+### Order of execution
+
+The life cycle events specified via environment variables is in addition to,
+and must not interfere with the life cycle operation order of `npm install` or
+other commands.
+
+In the following case:
+
+```
+# Build-time environment variables:
+$BP_NPM_PREINSTALL="env-preinstall-script"
+$BP_NPM_POSTINSTALL="env-postinstall-script"
+
+# package.json:
+{
+  "scripts" : {
+    "preinstall" : "packagejson-preinstall-script",
+    "install" : "packagejson-install-script",
+    "postinstall" : "packagejson-postinstall-script"
+    ...
+    ...
+  }
+}
+```
+
+The following must be the order of execution of events by the npm-install buildpack:
+```
+env-preinstall-script
+packagejson-preinstall-script
+packagejson-install-script
+packagejson-postinstall-script
+env-postinstall-script
+```
+
+The *env-&ast;-script*s should be explicitly executed by the buildpacks and the
+*packagejson-&ast;-script*s are implicitly executed as part of the buidpack's
+[npm-install process](https://docs.npmjs.com/cli/v6/using-npm/scripts#npm-install).
+
+### Open questions
+
+1. Web framework examples mostly demonstrate cases where a `postinstall` event
+   would be useful. What are some real world cases (if any) where `preinstall`
+   would be used?
+
+1. Do NPM and Yarn need separate env variables or should they read the same set
+   of env vars (e.g. `$BP_NODE_{PRE,POST}INSTALL`)?
```

Comment:
I like this configuration a little better. I think users will just look at their `package.json` to figure out what commands/scripts to run pre/post-install, so I think a single set of environment variables is easier.

Reply:
You mean the combined one? Which one would you prefer `$BP_NODE_*INSTALL` or `$BP_PACKAGEJSON_*INSTALL`?